File: benchmark-java/src/androidTest/java/com/shinkevich/benchmark_java/PiCpuLimitedBenchmark.java
package com.shinkevich.benchmark_java;

import android.util.Log;

import androidx.benchmark.junit4.BenchmarkRule;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;

import io.reactivex.rxjava3.schedulers.Schedulers;

@RunWith(value = Parameterized.class)
public class PiCpuLimitedBenchmark {
    private static final String TAG  = PiCpuLimitedBenchmark.class.getSimpleName();

    @Rule
    public BenchmarkRule benchmarkRule = new BenchmarkRule();
    @Parameterized.Parameter
    public int stepsNum;

    @Parameterized.Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {
                        {(int)Math.pow(10,4)},
                        {(int)Math.pow(10,5)},
                        {(int)Math.pow(10,6)},
                        {(int)Math.pow(10,7)},
                        {(int)Math.pow(10,8)},
                }
        );
    }

    @Test
    public void executorService_cpuLimitedThreads() throws InterruptedException, ExecutionException {
        int processorsNum = Runtime.getRuntime().availableProcessors();
        Log.d(TAG, "processors number: " + processorsNum);
        Tests.executorServiceTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Executors.newFixedThreadPool(processorsNum));
    }

    @Test
    public void rxJava_cpuLimitedThreads()  {
        int processorsNum = Runtime.getRuntime().availableProcessors();
        Log.d(TAG, "processors number: " + processorsNum);
        Tests.rxJavaTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Schedulers.computation());
    }
}

File: benchmark-kotlin/src/androidTest/java/com/shinkevich/benchmark_kotlin/PiCpuLimitedBenchmark.kt
package com.shinkevich.benchmark_kotlin

import android.util.Log
import androidx.benchmark.junit4.BenchmarkRule
import kotlinx.coroutines.Dispatchers
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized


@RunWith(value = Parameterized::class)
class PiCpuLimitedBenchmark(
    private val stepsNum: Int
) {
    @get:Rule
    var benchmarkRule = BenchmarkRule()

    @Test
    fun coroutines_cpuLimitedThreads() {
        val processorsNum = Runtime.getRuntime().availableProcessors()
        Log.d(TAG, "cpu processors: $processorsNum")
        Tests.coroutinesTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Dispatchers.Default.limitedParallelism(processorsNum))
    }

    @Test
    fun coroutines_cpuLimitedThreads_default() {
        val processorsNum = Runtime.getRuntime().availableProcessors()
        Log.d(TAG, "cpu processors: $processorsNum")
        Tests.coroutinesTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Dispatchers.Default)
    }

    @Test
    fun flow_cpuLimitedThreads(){
        val processorsNum = Runtime.getRuntime().availableProcessors()
        Log.d(TAG, "cpu processors: $processorsNum")
        Tests.flowTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Dispatchers.Default.limitedParallelism(processorsNum))
    }

    @Test
    fun flow_cpuLimitedThreads_default(){
        val processorsNum = Runtime.getRuntime().availableProcessors()
        Log.d(TAG, "cpu processors: $processorsNum")
        Tests.flowTest(benchmarkRule, stepsNum, Tests.getPrecision(stepsNum), processorsNum, Dispatchers.Default)
    }

    companion object {
        private val TAG = PiCpuLimitedBenchmark::class.java.simpleName

        @JvmStatic
        @Parameterized.Parameters
        fun data(): Iterable<Array<Any>> {
            return arrayListOf(
                arrayOf(Math.pow(10.0, 4.0).toInt()),
                arrayOf(Math.pow(10.0, 5.0).toInt()),
                arrayOf(Math.pow(10.0, 6.0).toInt()),
                arrayOf(Math.pow(10.0, 7.0).toInt()),
                arrayOf(Math.pow(10.0, 8.0).toInt()),
            ).toList() as Iterable<Array<Any>>
        }
    }
}